src/main/clojure/conexp/fca/factorization.clj

```diff
@@ -0,0 +1,940 @@
+        (let [X []]
+          (for [a A]
+            (into X (for [b B] (if (and (= a 1) (= b 1)) 1 0)))))))
+
+;; cartesian-product            
+(defn- cartesian-product
+  [a b]
+  (for [x a y b] [x y]))
+
+;; remove value from vector at index
+(defn- remove-indexed
+  [v n]
+  (into (subvec v 0 n) (subvec v (inc n))))
+
+;; counts false ones
+(defn count-false-ones
+  [V D]
+  (loop [i 0 j 0 c 0]
+    (if (>= i (count V))
+      c
+      (recur
+       (cond (> j (count (get V 0))) (inc i) :else i)
+       (cond (> j (count (get V 0))) 0 :else (inc j))
+       (cond (not= (get (get V i) j) (get (get D i) j)) (inc c) :else c)))))
+
+;; removes ones from vector
+(defn- remove-common-ones-vector
+  [A B]
+  (loop [i 0 out []]
+    (if (>= i (count A))
+      out
+      (recur
+       (inc i)
+       (if (= 1 (get A i)) (conj out 0) (conj out (get B i)))))))
+
+;; remove ones from matrix
+(defn- remove-common-ones-matrix
+  [A B]
+  (loop [i 0 out []]
+    (if (>= i (count A))
+      out
+      (recur
+       (inc i)
+       (conj out (remove-common-ones-vector (get A i) (get B i)))))))
+
+;; counts ones
+(defn count-one
+  [V]
+  (loop [i 0 c 0]
+    (if (> i (count V))
+      c
+      (recur (inc i) (cond (= (get V i) 1) (inc c) :else c)))))
+
+;; swaps the position of 2 elemnts in vector
+(defn swap-vector-position [v i1 i2]
+  (assoc v i2 (v i1) i1 (v i2)))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; evaluations used
+
+(defn hamming-distance
+  [A B]
+  (count (filter #{1} (map (fn [a b] (if (not= a b) 1 0)) A B))))
+
+(defn froebeniusnorm
+  [A]
+  (Math/sqrt (count (filter #{1} A))))
+
+(defn helper-eval
+  [ext-A set-ext-B]
+  (for [B set-ext-B]
+    (double (/ (count (clojure.set/intersection ext-A B)) (count (clojure.set/union ext-A B))))))
+
+(defn bicluster-match-score
+  [A B]
+  (let [ext-A (remove nil? (for [concept-A (concepts A)] (if (and (not-empty (first concept-A)) (not-empty (second concept-A))) (first concept-A))))
+        ext-B (remove nil? (for [concept-B (concepts B)] (if (and (not-empty (first concept-B)) (not-empty (second concept-B))) (first concept-B))))]
+        (* (/ 1 (count ext-A)) (clojure.core.reducers/reduce + (for [ext-a ext-A] (apply max (helper-eval ext-a ext-B)))))))
+
+(defn jaccard-index
+  "Computes the Jaccard index of two sets. This is |x ∩ y| / |x ∪ y|.
+  Returns 1 if both sets are empty. Taken from conexp.clj"
+  [x y]
+  (if (and (empty? x) (empty? y))
+    1
+    (/ (count (intersection x y)) (count (union x y)))))
+
+(defn false-incidences
+  "Retrieves a list of incidences which are present in context B but not in A"
+  [A B]
+  (difference (incidence-relation (normalize-context B)) (incidence-relation (normalize-context A))))
+
+(defn uncovered-incidences
+  "Retrieves a list of ubcovered incidences that A holds but are not present in context B"
+  [A B]
+  (difference (incidence-relation (normalize-context A)) (incidence-relation (normalize-context B))))
+
+(defn concept-ratio
```

Comment:
This name and defintion is weired, since the function does not know anything about concepts or contexts. Inline  (double ... )  and put a fitting comment there

Reply:
This was once used to determine the ratio of two different sets but was not actually used. Must've slipped when I cleaned the code